tests/test_language.py

```diff
@@ -236,6 +236,30 @@ def test_instruction_fusion_mscclpp():
     assert lowered_prgm.gpus[2].threadblocks[0].ops[2].inst == MscclppInstruction.signal
 
 
+def test_instruction_fusion_multi_deps_mscclpp():
+    topology = fully_connected(3)
+    collective = AllReduce(3, 1, True)
+    prgm = MSCCLPPProgram("allreduce", topology, collective, 1)
+    # last reduce_packet depends on put_packets(write after read)
+    # and first reduce_packet(rank 2 put data to rank 1, data put depends on first reduce_packet).
+    # In this case, we don't need to fuse the operations.
+    with prgm:
+        c0 = chunk(0, Buffer.input, 0)
+        c0.put_packet(1, "scratch", 0, sendtb=0)
+        c1s = chunk(1, "scratch", 0)
+        c1 = chunk(1, Buffer.input, 0)
+        c1 = c1.reduce_packet(c1s, recvtb=0)
+        c1.put_packet(0, "scratch", 0, sendtb=0)
+        c1.put_packet(2, "scratch", 0, sendtb=0)
+        c2 = chunk(2, Buffer.input, 0)
+        c2.put_packet(1, "scratch", 0, sendtb=0)
+        c1.reduce_packet(c1s, recvtb=0)
```

Comment:
In this case, the instruction dependency graph consists of the following 3 subgraphs:
1) put(0i to 1s) => reduce(1s to 1i) => put(2i to 1s) => reduce(1s to 1i)
2) reduce(1s to 1i) => put(1i to 0s)
3) reduce(1s to 1i) => put(1i to 2s)
There seems no cyclic dependency in this case?

Reply:
You miss the dependency between put(1i to 2s) => reduce(1s to 1i) (write after read)
Let's put all sub graph together. The whole graph would be:
```
put(0i to 1s) => reduce(1s to 1i) => put(2i to 1s) => reduce(1s to 1i)
                           | => put(1i to 0s)                      ^
                           |=>  put(1i to 2s)----------------------|
``` 
Let me update the comments